### PR TITLE
[GEOT-5667] Fix app-schema property substitution group support

### PR DESCRIPTION
**Pull request by braeckel**
https://osgeo-org.atlassian.net/browse/GEOT-5667

This pull request implements a GeoServer test for the GeoTools fix  https://github.com/geotools/geotools/pull/1502.  A new unit test was introduced with a single XML schema and properties file, similar to others in this directory.  This test case should fail when the related GeoTools fix is not integrated.

### CODE REVIEW

**Comment by bencaradocdavies** [comment]
Thanks, @braeckel. The failure is of course expected until the corresponding GeoTools PR is merged. I will review and test locally.

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/SubstitutionGroupCoverageWfsTest.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
+/* (c) 2014 Open Source Geospatial Foundation - all rights reserved
```
Comment (bencaradocdavies) [inline src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/SubstitutionGroupCoverageWfsTest.java]: Please also use new 2017 header here.

File: src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/SubstitutionGroupCoverageWfsTest.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        assertXpathEvaluatesTo("1", "/wfs:FeatureCollection/@numberReturned", doc);
+        assertXpathCount(1, "//test:DiscreteCoverage", doc);
+
+        String id = "a9274057-604d-427e-87f7-e6c9d846ceb5";
+        assertXpathEvaluatesTo(id, "(//test:DiscreteCoverage)[1]/@gml:id", doc);
+        assertXpathCount(1, "//test:DiscreteCoverage/test:domainSet", doc);
+        assertXpathCount(1, "//test:DiscreteCoverage/test:domainSet/test:DomainObject", doc);
+        assertXpathCount(1, "//test:DiscreteCoverage/test:domainSet/test:DomainObject/test:elements", doc);
+        assertXpathCount(2, "//test:DiscreteCoverage/test:domainSet/test:DomainObject/test:elements/*", doc);
+    }
```
Comment (bencaradocdavies) [inline src/extension/app-schema/app-schema-test/src/test/java/org/geoserver/test/SubstitutionGroupCoverageWfsTest.java]: You could an an assertion on the content of the ```test:elements```.

**Comment by bencaradocdavies** [comment]
@braeckel this PR adds new files. Has NCAR signed an OSGeo contributor agreement?
http://docs.geoserver.org/latest/en/developer/policies/committing.html

@simboss is the GeoServer project officer and can assist. @simboss, does OSGeo have an agreement from NCAR or UCAR that would cover contributions to GeoServer?

@braeckel the quickest solution is to get an authorised person at NCAR to sign this agreement:
http://www.osgeo.org/sites/osgeo.org/files/Page/corporate_contributor.txt
then send it in as instructed (plus a courtesy copy to @simboss (see developer list for his email)):
http://docs.geoserver.org/latest/en/developer/policies/committing.html

**Comment by bencaradocdavies** [comment]
@braeckel I am seeing some failures:

Running offline gs-app-schema-test tests I see:
```
Failed tests:   testGetFeature(org.geoserver.test.TimeSeriesWfsTest): expected:<1> but was:<0>
  testQuantityListSubset(org.geoserver.test.TimeSeriesWfsTest): expected:<1> but was:<0>
```

Running online tests against postgis I see this expand to:
```
Failed tests:   testTimePositionSubset(org.geoserver.test.TimeSeriesInlineWfsTest): expected:<1> but was:<0>
  testQuantityListSubset(org.geoserver.test.TimeSeriesInlineWfsTest): expected:<1> but was:<0>
  testGetFeature(org.geoserver.test.TimeSeriesInlineWfsTest): expected:<1> but was:<0>
  testGetFeature(org.geoserver.test.TimeSeriesWfsTest): expected:<1> but was:<0>

Tests in error: 
  testGetFeature(org.geoserver.test.SubstitutionGroupCoverageWfsTest): org.postgresql.util.PSQLException: ERROR: syntax error at or near "-"(..)
```
I expect that the error is caused by the hyphen in ```subgrp-coverage.properties```. The online tests convert your properties file into a database table. I am guessing that the table name is taken unquoted from the basename of the properties file name, and postgis does not permit an unquoted "-" in an identifier.

**Comment by braeckel** [comment]
@bencaradocdavies I have previously signed and submitted an individual contributor agreement.  I am working on the test failures and your other comments and will re-submit the PR when finished

**Comment by bencaradocdavies** [comment]
@braeckel good to hear that you have signed a contributor agreement. It can be a bureaucratic roadblock for some!

**Comment by bencaradocdavies** [comment]
@braeckel I am still watching this one. Outstanding fixes might be as single as renaming ```subgrp-coverage.properties``` to ```subgrpcoverage.properties``` and a fix in https://github.com/geotools/geotools/pull/1502 to avoid the timeseries case.

**Comment by bencaradocdavies** [comment]
I cherry-picked this PR onto master:

GEOT-5667 - added a GeoServer unit test for app schema substitution group behavior
https://github.com/geoserver/geoserver/commit/6238f36b2ad2c2e129ae1830d8456202db41b98c

I then applied the fixes I requested above:

[GEOT-5667] Rename subgrp-coverage.properties to subgrpcoverage.properties so that bare basename is a valid PostgreSQL identifier, fix copyright headers, add a test for test:elements, reorder subgrp.xml for schema-validity, and reformat
https://github.com/geoserver/geoserver/commit/b854305893c50fe81ef73a48f29c71b9ef9053bf

**Pull request closed without merging**